File: solutions/ruby/dominoes/1/dominoes.rb
# frozen_string_literal: true

# module Dominoes
module Dominoes
  def self.chain?(dominoes, chain = [])
    return chain.empty? || chain.first[0] == chain.last[1] if dominoes.empty?

    last = chain[-1]&.last
    dominoes.each_with_index.any? do |d, i|
      [d, d.reverse].any? do
        (last.nil? || _1[0] == last) && chain?(dominoes[0...i] + dominoes[i + 1..], chain + [_1])
      end
    end
  end
end

File: solutions/clojure/dominoes/1/src/dominoes.clj
(ns dominoes)

(defn can-chain? [dominoes]
  (if-let [[[chain-start open-end] & unplaced-dominoes] (seq dominoes)]
    (letfn [(chainable? [open-end unplaced-dominoes]
              (if (empty? unplaced-dominoes)
                (= chain-start open-end)
                (boolean
                 (some (fn [[index [left right]]]
                         (let [remaining-dominoes
                               (concat (take index unplaced-dominoes)
                                       (drop (inc index) unplaced-dominoes))]
                           (cond
                             (= open-end left) (chainable? right remaining-dominoes)
                             (= open-end right) (chainable? left remaining-dominoes))))
                       (map-indexed vector unplaced-dominoes)))))]
      (chainable? open-end unplaced-dominoes))
    true))
